### ISSUE TITLE
peekAll after edit then unload leaves lingering record

### ISSUE THREAD

**Issue by hashishaw** [emberjs/data#8863]
This issue might be related to #8299 (which the test is modeled after, thank you 😁)

### Reproduction

Here is a test outlining the issue:

```js
import { run } from '@ember/runloop';
import Model, { attr } from '@ember-data/model';
import { setupTest } from 'ember-qunit';
import { module, test } from 'qunit';
import Adapter from '@ember/test/adapter';

module('Unit | Model | edit record then unloadAll', function (hooks) {
  setupTest(hooks);

  hooks.beforeEach(function () {
    const Company = Model.extend({
      name: attr('string'),
    });

    const CompanyAdapter = Adapter.extend({
      updateRecord: () => {
        return undefined;
      },
    });

    this.owner.register('model:company', Company);
    this.owner.register('adapter:company', CompanyAdapter);
    this.store = this.owner.lookup('service:store');
  });

  test('edit then unload does not remove all records', async function (assert) {
    this.store.push({
      data: [
        {
          id: '1',
          type: 'company',
          attributes: {
            name: 'ACME',
          },
        },
        {
          id: '2',
          type: 'company',
          attributes: {
            name: 'EMCA',
          },
        },
      ],
    });
    assert.strictEqual(this.store.peekAll('company').length, 2, '2 companies loaded');
    const editRecord = this.store.peekRecord('company', '1');
    editRecord.name = 'Rebrand';
    await editRecord.save();
    assert.false(editRecord.hasDirtyAttributes, 'edit record does not have dirty attrs after save');
    run(() => {
      this.store.unloadAll('company');
    });

    // fails b/c value is 1 instead of 0
    assert.strictEqual(this.store.peekAll('company').length, 0, 'peekAll 0 - companies unloaded');
    // fails with Assertion Failed: Expected to receive a stable Identifier to subscribe to
    assert.strictEqual(
      this.store.peekAll('company').toArray().length,
      0,
      'peekAll array 0 - companies unloaded'
    );
  });
});
```

### Description

After loading a list of records, if we edit a record and then unloadAll there is a lingering record in `peekAll` with a `null` ID.  I would expect that unloading after editing a record would still unload all the records of that type. 

### Versions

This scenario worked in our previous version of ember-data@4.5.0

Current version: 
```cli
Ember: 4.12.3
Ember Data: 4.11.3
```

Please let me know if I can provide any additional information, and would appreciate a workaround for the time being if there is one. Thank you!

**Comment by runspired** [comment]
does this still reproduce in 4.12+?

My intuition is that if there's a bug here its that the RecordArray has not needed to flush its notifications queue at the point you call unloadAll. If that is the case, calling peekAll prior to unloadAll is a workaround.

**Comment by runspired** [comment]
looks like this issue was fixed in 4.12+ so going to close, thanks for the report and test!

**Comment by hashishaw** [comment]
Thanks for the quick response! 

> If that is the case, calling peekAll prior to unloadAll is a workaround.
this is a funny one, if I just peekAll before running unload it has the same behavior. But if I get the length of peekAll it works. 🤷 

In any case, glad this is resolved in 4.12, will look into updating to that version. Thanks again!
